Artesian/_ClientsExecutor/RequestExecutor.py

```diff
@@ -251,4 +252,20 @@ def __init__(self, last_attempt):
         self.last_attempt = last_attempt
 
     def __str__(self):
-        return "RetryError[{0}]".format(self.last_attempt)    
\ No newline at end of file
+        return "RetryError[{0}]".format(self.last_attempt)    
+
+
+def createSemaphore(parallelism):
+    """
+    Wrapper around asyncio semaphore.
+    Ensures that there is an even loop available.
+    An event loop may not be available if the sdk is not run in the main event loop
+    """
+    try:
+        asyncio.get_event_loop()
+    except RuntimeError as ex:
```

Comment:
This looks strange. 

get_event_loop() is not supposed to fail per documentation (https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_event_loop) but exactly to create a new event loop and set it.

Reply:
The get_event_loop() won't fail if its called from the main os thread. The issue is that a user is calling the sdk from within an event loop context, get_event_loop seems to fail if it is inside an event loop context that does not have it's own even loop set. So I try catch on this scenario to ensure an event loop is present

Reply:
I still think is the wrong place to handle. This private code is supposed to be always in 'async' context, ensured by the syncronous execute(). I think this 'loop' creation in a non-main thread should be done there.


The _exec() looks the place where the current library should handle this. Is the compatibility layer between the Async nature of the library and the Sync execution.

There is also a new https://docs.python.org/3/library/asyncio-task.html#asyncio.run that should take care of that. What do you think?

```py
    def _exec(self, urls):
        loop = asyncio.get_event_loop()
        rr = loop.run_until_complete(self._execAsync(urls))
        return rr
```

If is already an async context, isn't responsability of the caller to be sure the loop is configured instead?

Reply:
"If is already an async context, isn't responsability of the caller to be sure the loop is configured instead?"
I don't think so. They are calling what they expect is a synchronous function the fact we use async internally is opaque to the caller. 

I think it makes more sense to add it to _exec. I put it there originally but the semaphore also needed it. I'll try add it to exec with the lazy evaluated semaphore